File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第3章：空间变换/3.2-投影变换/perspective.py
# -*- coding: utf-8 -*-
import numpy as np
import cv2
import sys
#主函数
if __name__ == "__main__":
    if len(sys.argv)>1:
        image = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usage: python warpPercpective.py image"
    #原图的宽高
    h,w = image.shape
    #将原图的四个顶点投影到一个不规则的四边形中
    src=np.array([[0,0],[w-1,0],[0,h-1],[w-1,h-1]],np.float32);
    dst=np.array([[20,50],[w/2,150],[50,h/2],[w-40,h-40]],np.float32)
    #计算投影矩阵
    p=cv2.getPerspectiveTransform(src,dst)
    #利用计算出的投影矩阵进行头像的投影变换
    r = cv2.warpPerspective(image,p,(w,h),borderValue=0)
    #显示原图和投影效果
    cv2.imshow("image",image)
    cv2.imshow("warpPerspective",r)
    cv2.imwrite("p.jpg",r)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第3章：空间变换/3.2-投影变换/perspective.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/imgproc/imgproc.hpp>
#include<opencv2/highgui/highgui.hpp>
using namespace cv;
#include<iostream>
using namespace std;
Mat I;//原图
Mat pI;//投影变换后的图
Point2f IPoint,pIPoint;
int i = 0,j = 0;
Point2f src[4];//存储原坐标
Point2f dst[4];//存储对应变换的坐标
//通过以下鼠标事件，在原图中取四个坐标
void mouse_I(int event, int x, int y, int flags, void *param)
{
	switch (event)
	{
	case CV_EVENT_LBUTTONDOWN:
		//记录坐标
		IPoint = Point2f(x, y);
		break;
	case CV_EVENT_LBUTTONUP:
		src[i] = IPoint;
		circle(I, src[i], 7, Scalar(0),3);//标记
		i += 1;
		break;
	default:
		break;
	}
}
//通过以下鼠标事件，要输出的图中取四个坐标
void mouse_pI(int event, int x, int y, int flags, void *param)
{
	switch (event)
	{
	case CV_EVENT_LBUTTONDOWN:
		//记录坐标
		pIPoint = Point2f(x, y);
		break;
	case CV_EVENT_LBUTTONUP:
		dst[j] = pIPoint;
		circle(pI, dst[j], 7, Scalar(0), 3);//标记
		j += 1;
		break;
	default:
		break;
	}
}
int main(int argc, char*argv[])
{
	//输入原图
	I = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!I.data)
		return 0;
	//输出图像
	pI = 200 * Mat::ones(I.size(), CV_8UC1);
	//在原图窗口上，定义鼠标事件
	namedWindow("I", 1);
	setMouseCallback("I", mouse_I, NULL);
	//在输出窗口上，定义鼠标事件
	namedWindow("pI", 1);
	setMouseCallback("pI", mouse_pI, NULL);
	imshow("I", I);
	imshow("pI", pI);
	while(!(i == 4 && j == 4))
	{
		imshow("I", I);
		imshow("pI", pI);
		if (waitKey(50) == 'q')
			break;
	}
	imshow("I", I);
	imshow("pI", pI);
	imwrite("I.jpg", I);
	imwrite("pI.jpg", pI);
	//移除鼠标事件 
	setMouseCallback("I", NULL, NULL);
	setMouseCallback("pI", NULL, NULL);
	//计算投影变换矩阵
	Mat p = getPerspectiveTransform(src, dst);
	//投影变换
	Mat result;
	warpPerspective(I, result, p, pI.size());
	imshow("投影后的效果", result);
	imwrite("result.jpg", result);
	waitKey(0);
	return 0;
}
